doc/design/sql_program_dependency_analyze.md

```diff
@@ -0,0 +1,161 @@
+TO TRAIN DNNClassifier
+WITH model.n_classes=2, model.hidden_units=[4096,1024,256]
+LABEL 'label'
+INTO my_dnn_model;
+-- train another model using XGBoost
+SELECT * FROM training_data
+TO TRAIN xgboost.gbtree
+WITH objective="binary:logistic"
+LABEL 'label'
+INTO my_xgb_model;
+-- generate explain results for both model for analyze
+SELECT * FROM training_data ORDER BY RAND() LIMIT 10000
+TO EXPLAIN my_dnn_model;
+
+SELECT * FROM training_data ORDER BY RAND() LIMIT 10000
+TO EXPLAIN my_xgb_model;
+
+-- clean up training table
+DROP TABLE training_data;
+```
+
+It's simple to figure out that we can run the two training statement concurrently, and run the
+corresponding explain statement when one training statement finishes. The execution flow should
+look like:
+
+<p align="center">
+<img src="figures/deps_flow_graph.png">
+</p>
+
+## Analyze the SQL Program Execution Dependency
+
+By analyzing the SQL program, we are able to figure out which statements can be executed
+concurrently. Since the SQL statements are often submitted to a distributed cluster like
+Hive or MaxCompute. Running the SQL statements concurrently can make full use of the cluster
+computing resource, shrink the execution time.
+
+To achieve this, we need to analyze the hole SQL program and find out what tables are manipulated
+by each statement. One statement can read, write or do both on some tables, while other statements
+can also read, write the same table, in the above case, statement 1 is reading table `user_visits`
+and writing table `training_data`, statement 2,3,4,5 are reading table `training_data`, so they must
+start only when statement 1 is finished.
+
+We can do this by following below steps:
+
+1. Parse every SQL statement, return a map recording read/write type of the tables that the current
+   statement is manipulating. We'll get a list of this map in order, representing each SQL statement.
+   This will be done in the third-party parsers.
+2. Parse the SQLFlow extended SQL if there are any, return a similar map recording read/write type
+   of the models of the current statement (in our Go parser).
+3. Construct a graph using the following rules:
+    1. The statements are nodes in the graph, they should run in the original order.
+    2. If one statement reads a table/model, then construct one graph edge from statements that write
+       the table to the current statement.
+4. Execute this graph by submitting the graph as an Argo/Tekton workflow, each step is one single
+   statement.
+
+## Analyzing Database Context
+
+Some SQL program have `USE` clause which defines the database the following SQL statements are using.
+
+```SQL
+USE db;
+CREATE TABLE my_table AS ...
+SELECT * FROM my_table TO TRAIN ...
+USE db2;
+SELECT * FROM my_table_in_db2 TO TRAIN ...
+```
+
+We need to go over the SQL program and get the last `USE` for each SQL statement, and modify the table
+name with out prefix `db.` to `db.table` for later dependency analyzing.
```

Comment:
Do we need to change the external parser interface to identify `USE`-statements from "normal" statements? If so, we need a section describing how.

Reply:
No, updating current external parser grammar seems a lot of work, so we may use a simple way using regex matching.